### DISCUSSION TITLE
RandomIC

### DISCUSSION THREAD

**Discussion by manikumd22** [idaholab/moose#28666] [Q&A Modules: Solid mechanics] [answered]
Hi @GiudGiud 

I am working on generating the RandomIC. I want to assign each quad point a different Youngs modulus, by multiplying with RandomIC. I used this way to do that, but I'm getting an error. 

/storage/work/mvk6158/moose_directories/myApp/new_mesh1.i:33.7: invalid float syntax for parameter: Materials/elasticity1/youngs_modulus=10000*c

#Tensor Mechanics tutorial: the basics
#Step 1, part 1
#2D simulation of uniaxial tension with linear elasticity
```
[GlobalParams]
    displacements = 'disp_x disp_y'
[]
  
[Mesh]
    file = manik_7.msh
[]

[ICs]
  [./cIC]
    type = RandomIC
    variable = c
    min = 0.01
    max = 0.2
  [../]
[]
  
[Modules/TensorMechanics/Master]
  [all]
  add_variables = true
  strain = small
  generate_output = 'vonmises_stress stress_xx stress_yy strain_xx strain_yy'
  []
[]
  
[Materials]
  [elasticity1]
      type = ComputeIsotropicElasticityTensor
      youngs_modulus = '10000*c'
      poissons_ratio = 0.1
  []
  [./stress]
      type = ComputeLinearElasticStress
  [../]
[]
  
[BCs]
[./left]
    type = DirichletBC
    variable = disp_x
    boundary = 'top bottom'
    value = 0.0
[../]
[./bottom]
    type = DirichletBC
    variable = disp_y
    boundary = 'top'
    value = 1
[../]
[./top]
    type = DirichletBC
    variable = disp_y
    boundary = 'bottom'
    value = 0
[../]
[]
  
[Preconditioning]
  [./SMP]
      type = SMP
      full = true
  [../]
[]


[Executioner]
  type = Steady
  solve_type = 'NEWTON'
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -sub_pc_type -pc_asm_overlap -ksp_gmres_restart'
  petsc_options_value = 'asm lu 1 101'
[]
  
[Outputs]
  csv = false
  exodus = true
  file_base = ./inputfiles/manikcode_ir_out_new12
[]
```

**Comment by GiudGiud** [comment]
Hello

This error message

> /storage/work/mvk6158/moose_directories/myApp/new_mesh1.i:33.7: invalid float syntax for parameter: Materials/elasticity1/youngs_modulus=10000*c

tells you it's expecting a float (a single number) but you are trying to pass some sort of parsed expression involving a variable.
You can't do that, but you can use a different material, namely this one
https://mooseframework.inl.gov/source/materials/ADComputeVariableIsotropicElasticityTensor.html

this one expects material properties for the young modulus. You could use the ADParsedMaterial to perform this 10000 * c operation and set the material property

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Is this how should I define it?

I defined this way and getting error?
```
[31m
*** ERROR ***
The following error occurred in the Problem 'MOOSE Problem' of type FEProblem.

Variable 'c' requested in initial condition 'cIC' does not exist.[39m

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0

[ICs]
  [./cIC]
    type = RandomIC
    variable = c
    min = 0.01
    max = 0.2
  [../]
[]
  
  
[Materials]
  [compute_stress]
    type = ComputeNeoHookeanStress
    lambda = E_phys
    mu = 6700
  []
  [E_phys]
    type = ParsedMaterial
    expression = "4000*c"
    coupled_variables = 'mat_den'
    property_name = E_phys
  []
  [./stress]
      type = ComputeLinearElasticStress
  [../]
[]
```

**Reply by GiudGiud** [reply]
```
[E_phys]
type = ParsedMaterial
expression = "4000*c"
coupled_variables = 'c'
property_name = E_phys
[]
```

**Reply by GiudGiud** [reply]
and also add 
```
[AuxVariables]
  [c]
  []
[]
```
to create the variable

**Reply by manikumd22** [reply]
Hi @GiudGiud 

This has worked, but I used for this expression

[Materials]
  [compute_stress]
    type = Compute_NeoHookean
    lambda = 4000
    mu = 6700
    quad_points = '0.01761040545043226 0.06572292788290107 0.10094373878376559 0.14905626121623441 0.18427707211709893 0.23238959454956776 0.26761040545043224 0.31572292788290107 0.3509437387837656 0.3990562612162345 0.434277072117099 0.4823895945495678 0.5176104054504322  0.5657229278829011 0.6009437387837656 0.6490562612162344 0.6842770721170989 0.7323895945495679 0.7676104054504324 0.8157229278829011 0.8509437387837657 0.8990562612162345 0.934277072117099 0.9823895945495678'
    no_quad_points = 24

    stiff = E_phys
  []
  [E_phys]
    type = ParsedMaterial
    expression = "0*c 0.04347826*c 0.08695652*c 0.13043478*c 0.17391304*c 0.2173913*c 0.26086957*c 0.30434783*c 0.34782609*c 0.39130435*c 0.43478261*c 0.47826087*c 0.52173913*c 0.56521739*c 0.60869565*c 0.65217391*c 0.69565217*c 0.73913043*c 0.7826087*c 0.82608696*c 0.86956522*c 0.91304348*c 0.95652174*c 1*c"
    coupled_variables = 'c'
    property_name = E_phys
  []

It showed some error

[31m
*** ERROR ***
/storage/work/mvk6158/moose_directories/myApp/parsed_mat.i:211.5: cannot convert field 'Materials/compute_stress/stiff' value 'E_phys' to float
[39m


I wanted to  use this c at every gauss point in this way

for (int p = 0; p < 24; ++p) {     
        if (MooseUtils::absoluteFuzzyEqual(_q_point[_qp](0),  _quad_points[p]))
        {   
        RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
        _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
                _mu[_qp] * RankTwoTensor::Identity();
        _C[_qp] =
            -2 * (_stiff[_qp]*_lambda[_qp] * log(_F[_qp].det()) - _stiff[_qp]*_mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
            _stiff[_qp]*_lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
        }
    }

**Reply by GiudGiud** [reply]
the Compute_NeoHookean material also expects a Real for stiff. If you want to use a material property, you ll need to modify that material to use that instead of the parameter

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Got it thank!


One thing I would like to know about the seed() in this,

If I want to define seed(0.2), seed(0.5), seed(0.8).

What does it indicate and how to use in this ICs

[ICs]
  [./cIC]
    type = RandomIC
    variable = c
    min = 0
    max = 1
  [../]
[]

**Reply by GiudGiud** [reply]
the seed is used to generate a different set of random numbers.

**Reply by manikumd22** [reply]
Hi @GiudGiud,
With this, if I want an average of 0.5, 0.1, and 0.9. What changes should I make to get that average with this ICx


[ICs]
[./cIC]
type = RandomIC
variable = c
min = 0
max = 1
[../]
[]

**Reply by GiudGiud** [reply]
you ll need to create a new object. The randomIC is not designed to obey an average value

The new object could use an average parameter and a distribution parameter to handle the randomness
https://mooseframework.inl.gov/syntax/Distributions/index.html

**Reply by manikumd22** [reply]
Hi @GiudGiud ,


I tried this way but got the error

*** ERROR ***
A 'DistributionIC' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.[39m


[Distributions]
  [./beta_c]  # Change this to 'beta_c' to match the ICs block
    type = Beta
    alpha = 1.9591684282393695
    beta = 3.1004701483699475
  [../]
[]

[ICs]
  [./cIC]
    type = DistributionIC
    variable = c
    distribution = beta_c  # Now correctly linked to the Beta distribution
  [../]
[]

Is this syntax correct for using the beta distribution for my problem?
Do I need to change the rest of the code?

**Reply by GiudGiud** [reply]
you need to code this object. where did you see a distributionIC?

**Reply by manikumd22** [reply]
Hi @GiudGiud ,

Can you please help me, I don't know how to code this object.

for distribtuionIC, I was just trying some iterations

**Reply by GiudGiud** [reply]
turns out you can specify the distribution directly in randomIC

so use a RandomIC with a distribution parameter.
And in the distribution, select one that has the desired average and spread. you can tune a Guassian for example
https://mooseframework.inl.gov/syntax/Distributions/index.html

**Reply by manikumd22** [reply]
Hi @GiudGiud 

[Distributions]
  [./beta_c]  # Change this to 'beta_c' to match the ICs block
    type = Beta
    alpha = 3.0
    beta = 3.0
  [../]
[]

[ICs]
  [./cIC]
    type = RandomIC
    variable = c
    distribution = beta_c  # Now correctly linked to the Beta distribution
  [../]
[]


It is working but the solution is skipped, 

Time Step 1, time = 0.05, dt = 0.05
[32m Solve Skipped![39m

**Reply by GiudGiud** [reply]
do you have solve = false in the input file somewhere?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I solved that error

Thanks for the helping me out

**Reply by manikumd22** [reply]
Hi @GiudGiud 
Just have one more question,

Is it possible to assign the c as linear variation?

**Reply by GiudGiud** [reply]
So not random?

Or a linear random distribution?

Or the random number is used as a linear coefficient?

**Reply by manikumd22** [reply, marked as answer]
Hi @GiudGiud 

I solved that

Thanks again for helping me
